lib/Sema/TypeCheckConcurrency.cpp

```diff
@@ -1969,6 +2152,72 @@ namespace {
             ConcurrentReferenceKind::CrossActor);
       }
 
+      case ActorIsolationRestriction::DistributedActor: {
+        /// mark for later diagnostics that we have we're in a distributed actor.
+        isDistributedActor = true;
+
+        // distributed actor isolation is more strict;
+        // we do not allow any property access, or synchronous access at all.
+
+        bool continueToCheckingLocalIsolation = false;
+        // Must reference distributed actor-isolated state on 'self'.
+        auto *selfVar = getReferencedSelf(base);
+        if (!selfVar) {
+          // invocation on not-'self', is only okey if this is a distributed func
+
+          if (auto func = dyn_cast<FuncDecl>(member)) {
+            if (!func->isDistributed()) {
+              ctx.Diags.diagnose(memberLoc, diag::distributed_actor_isolated_method);
+              // TODO: offer a fixit to add 'distributed' on the member; how to test fixits? See also https://github.com/apple/swift/pull/35930/files
+              noteIsolatedActorMember(member, context, isDistributedActor);
+              return true;
+            }
+
+            assert(func->isDistributed());
+
+            if (!func->hasAsync())
+              tryMarkImplicitlyAsync(memberLoc, memberRef, context);
+
+            tryMarkImplicitlyThrows(memberLoc, memberRef, context);
+
+            // TODO: mark implicitly async here?
+            // TODO: we don't really need to do anythign with the result, dont get it?
+
+            // distributed func reference, that passes all checks, great!
+            continueToCheckingLocalIsolation = true;
+          } // end FuncDecl
+
+          if (!continueToCheckingLocalIsolation) {
+            // it wasn't a function (including a distributed function),
+            // so we need to perform some more checks
+            if (auto var = dyn_cast<VarDecl>(member)) {
+              // @_distributedActorIndependent decls are accessible always,
+              // regardless of distributed actor-isolation; e.g. actorAddress
+              if (member->getAttrs().hasAttribute<DistributedActorIndependentAttr>())
+                return false;
+
+              // otherwise, no other properties are accessible on a distributed actor
+              if (!continueToCheckingLocalIsolation) {
+                ctx.Diags.diagnose(
+                    memberLoc, diag::distributed_actor_isolated_non_self_reference,
+                    member->getDescriptiveKind(),
+                    member->getName());
+                noteIsolatedActorMember(member, context, isDistributedActor);
+                return true;
+              }
+            }
+
+            // TODO: would have to also consider subscripts and other things
+          }
+        } // end !selfVar
+
+        return false;
```

Comment:
There might be more checking needed here, depending on what kinds of members are allowed in a `distributed actor`. For example, for normal actors, even if the access is through `self`, that doesn't mean the member has the same actor-isolation as `self`:

```
actor A {
  @MainActor func f() {}
  func g() {
    self.f() // self-accessed member, but the actor-isolation doesn't match up,
        // so you stil need to mark it as implicitly-async and do ConcurrentValue checking.
  }
}
```

I'm not sure if members of a distributed actor will be allowed to have different actor-isolation, but I just wanted to point this out so that you're more certain that you got this checking right 😄

Reply:
Oh that's a great case, thanks I'll think about it some more 👍